First prototype of native pytest plugin for launch based tests

ivanpauno opened the pull request:
Some early discussion can be found here: https://gist.github.com/ivanpauno/761a7cbca83335ea081512f3494304ba.

I'm planning to add support for shutdown tests.
There are also a lot of other details that can be discussed.

hidmic commented on launch_testing/launch_testing/pytest/fixture.py:
```diff
@@ -0,0 +1,88 @@
+                ):
+                    mod_locals[name] = getter(scope)
+            else:
+                mod_locals[name] = getter(scope)
```
@ivanpauno I think I understand the override-able fixture magic, but doesn't this imply fixtures of different scopes cannot coexist in the same source file? I mean, you could have 3 tests using a module scope fixture and 3 tests using a function scope fixture and there wouldn't be any overlap, would it?

ivanpauno replied:
If you have one launch_test fixture with `class` scope and one with `function` scope, the launch_service function will be class scoped.

This is required because of the following issue:

```python3

@pytest.fixture(scope='module')
def foo():
  ...

@pytest.fixture()
def bar(foo):
  ...
```

In that case you get a "scope mismatch error", because you cannot access a fixture with a lower scope from another fixture.

The fact that the launch_service/event_loop is automatically redefined with a larger scope only means that it might be shared by both launch_test fixtures, which is not necessarly a problem.
We could actually create one launch_service per launch_test fixture, and don't use a fixture for that at all.
For the event_loop though, you can only have one active event loop in a thread, so it would be an issue anyway.

Maybe in the future pytest solves this issue, see https://github.com/pytest-dev/pytest/issues/6101.

hidmic replied:
> In that case you get a "scope mismatch error", because you cannot access a fixture with a lower scope from another fixture.

Aye, the `launch_service` fixture scope has to be at least as large as the largest scope of all launch fixtures found (BTW isn't that example reversed?).

> The fact that the launch_service/event_loop is automatically redefined with a larger scope only means that it might be shared by both launch_test fixtures, which is not necessarly a problem.

I see two problems.  Unrelated larger scope fixtures creeping into tests with smaller scope fixtures, and a large scope launch fixture being shutdown (which IIRC is a terminal state) by a smaller scope fixture upon finalization. The first one is not really a problem if we enforce a 1-to-1 relationship between tests and launch fixtures. The second one is.

> We could actually create one launch_service per launch_test fixture, and don't use a fixture for that at all.

Having separate instances is probably easiest, yeah. Alternatively, we can make `LaunchService` shutdowns not a terminal state (i.e. allow re-running after a shutdown).

ivanpauno replied:
> (BTW isn't that example reversed?).

lol yes, fixed.

> and a large scope launch fixture being shutdown (which IIRC is a terminal state) by a smaller scope fixture upon finalization

Yes, I think we can detect that case and warn the user though.

hidmic reviewed (commented):
First pass, great stuff!

wjwwood commented on launch/launch/launch_service.py:
```diff
@@ -413,3 +413,7 @@ def shutdown(self, force_sync=False) -> Optional[Coroutine]:
         """Getter for context."""
         return self.__context
+
+    @property
+    def event_loop(self):
+        return self.__loop_from_run_thread
```
nitpick: Not sure if it's needed (like if a linter will complain) but there's no docstring here.

ivanpauno replied:
fixed https://github.com/ros2/launch/pull/528/commits/38098a6c43261ec60be9ed6a38930c7d57144d26

wjwwood commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,258 @@
+    """Inject launch_testing marker documentation."""
+    config.addinivalue_line(
+        'markers',
+        'launch_testing: '
```
I don't have an opinion about this. `launch_test` sounds natural, but the package name is `launch_testing` which would be consistent. 🤷

ivanpauno replied:
> @ivanpauno meta: I subjectively prefer launch_test for a marker.

We [already have one](https://github.com/ros2/launch/blob/8a7649de4d65d13e24f176f2005917a9ba3061a0/launch_testing/launch_testing/pytest/hooks.py#L185) named like that, we can probably avoid colisions based on the arguments, but I would rather use a different marker name.

hidmic replied:
Argh, true. `launch_pytest` :sweat_smile: ?  `pytest.mark.launch_testing` would be consistent with e.g. `pytest.mark.asyncio`, but I find both a little awkward. Anyhow, not a big concern really.

wjwwood commented on launch_testing/test/launch_testing/new_hooks/test_function_scope.py:
```diff
@@ -0,0 +1,33 @@
+
+
+@launch_testing.pytest.fixture()
+def ld():
```
nitpick: `ld` isn't an ideal name, use `launch_description` if this is kind of a "callable object" or `get_launch_description` if a function name (use a verb for a function name)

ivanpauno replied:
> nitpick: ld isn't an ideal name, use launch_description if this is kind of a "callable object" or get_launch_description if a function name (use a verb for a function name)

`launch_description` sounds fine, pytest.fixtures names are generally a noun phrase and not a verb, because that's also the name of the argument automatically passed to tests.

ivanpauno replied:
fixed

wjwwood reviewed (commented):
I left a few comments, but honestly the pytest stuff is a bit over my head (opaque without some special knowledge it seems).

hidmic commented:
On a related note, would it make sense to roll out a separate `launch_pytest` package? Imports would be shorter and it's easier to draw a line between this new thing and `launch_testing`. WDYT @ivanpauno @wjwwood @jacobperron ?

ivanpauno commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,258 @@
+        item = pytest.Function.from_parent(collector, name=name)  # To reload keywords.
+
+        if 'launch_testing' in item.keywords:
+            return list(collector._genfunctions(name, obj))
```
why?
The [pytest_pycollect_makeitem()](https://docs.pytest.org/en/6.2.x/reference.html#pytest.hookspec.pytest_pycollect_makeitem) plugin [stops at the first not `None` result](https://docs.pytest.org/en/6.2.x/writing_plugins.html#firstresult).

pytest-asyncio uses a similar hook.

hidmic replied:
Ooh, right, `pytest` hooks are associated to N functions. Nevermind.

ivanpauno commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,258 @@
+        ls.include_launch_description(ld)
+        run_async_task = event_loop.create_task(ls.run_async(
+            # TODO(ivanpauno): maybe this could be configurable (?)
+            shutdown_when_idle=True
```
This code is only executed for `@launch_testing.pytest.fixture()` fixtures, and not for anything else.
If you want to access the launch service fixture directly and you don't have a launch_testing fixture, it won't be a problem.

Maybe I'm not understanding your comment though.
We could also have a `shutdown_when_idle` argument in `@launch_testing.pytest.fixture()`

hidmic replied:
> This code is only executed for `@launch_testing.pytest.fixture()` fixtures, and not for anything else.

Ahh, good point.  

> If you want to access the launch service fixture directly and you don't have a launch_testing fixture, it won't be a problem.

Hmm, what's putting the launch service to run in the event loop in that case? That's a prerequisite for `wrap_coroutine` to work.

> We could also have a shutdown_when_idle argument in @launch_testing.pytest.fixture()

Yeah, but it's less obvious to me why you'd want to do that in the general case (besides adding it because it's easy).

ivanpauno replied:
> Hmm, what's putting the launch service to run in the event loop in that case? That's a prerequisite for wrap_coroutine to work.

We're only using `wrap_coroutine` in tests marked with `pytest.mark.launch_testing`, which requires a fixture.
For other async tests that take a launch_service fixture directly, I think the magic that `pytest-asyncio` uses would be enough (those tests will require a `pytest.mark.asyncio` decorator).
We can certaninly test a case like that.

hidmic replied:
> We can certaninly test a case like that.

Yeah, I'd like to see how it all plays together (fixtures with different scopes, launch tests, asyncio tests that use launch_service).

ivanpauno replied:
> Yeah, but it's less obvious to me why you'd want to do that in the general case (besides adding it because it's easy).

Couldn't this replace the `KeepAliveProc()` thing?
There's really no need to launch a process to keep the launch service alive.

ivanpauno replied:
How does that currently work in launch_testing by the way?

Do we stop running tests when the launch_service is idle and mark them as failed?
Should do we doing something similar here?
The code here is only ensuring that the launch service starts running before the tests and it's cleaned up before the shutdown tests, but it doesn't ensure it isn't idle before finishing running the (not shutdown) tests.

hidmic replied:
Looking at [test_runner.py](https://github.com/ros2/launch/blob/90a245b493023d4d979d0adf8c6583083e9e33ba/launch_testing/launch_testing/test_runner.py#L163), if the launch service running task finishes before the tests, we fail. 

> Couldn't this replace the KeepAliveProc() thing?

BTW https://github.com/ros2/launch/blob/90a245b493023d4d979d0adf8c6583083e9e33ba/launch_testing/launch_testing/test_runner.py#L159 :sweat_smile:

ivanpauno replied:
> Looking at test_runner.py, if the launch service running task finishes before the tests, we fail.

Okay.
I'm not sure if keeping that behavior makes sense though.

> :sweat_smile:

It's even more fun knowing that the process launched by `KeepAliveProc()` was using 100% of one cpu.

hidmic replied:
> It's even more fun knowing that the process launched by KeepAliveProc() was using 100% of one cpu.

Oh boy :see_no_evil:

ivanpauno commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,258 @@
+        raise RuntimeError(
+            'launch_pytest fixtures must return a LaunchDescription '
+            'containing a ReadyToTest action'
+        )
```
Yes, appending one to the launch description actions sounds like a reasonable default.

ivanpauno commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,258 @@
+                f'{get_error_context_from_obj(item.obj)}'
+            )
+        fixturename = marker.kwargs['fixture'].__name__
+        if fixturename in item.fixturenames:
+            item.fixturenames.remove(fixturename)
+        item.fixturenames.insert(0, fixturename)
+        if 'launch_service' in item.fixturenames:
+            item.fixturenames.remove('launch_service')
+        item.fixturenames.insert(0, 'launch_service')
```
I don't think they need to go first, I'm just avoiding to duplicate the item.

hidmic replied:
Hmm, would: 
```suggestion
        if fixturename not in item.fixturenames:
            item.fixturenames.append(fixturename)
        if 'launch_service' not in item.fixturenames:
            item.fixturenames.append('launch_service')
```
achieve the same?

ivanpauno replied:
I think so, yes

ivanpauno replied:
this code was completely modified in https://github.com/ros2/launch/pull/528/commits/46ba3b249c5f23869d750816bba35c67f09d4ce1

ivanpauno commented:
> On a related note, would it make sense to roll out a separate launch_pytest package? Imports would be shorter and it's easier to draw a line between this new thing and launch_testing. WDYT @ivanpauno @wjwwood @jacobperron ?

The imports wouldn't really be much shorter.
Having a separate package sounds fine to me, I actually started that way and then integrated the changes in launch_testing.

There're a few things that I can reuse from the current launch_testing though, like the `ReadyToTest` action.
I'm not sure if we can also take advantage of some of the existent tools to e.g. launch a new process, capture process stdout, etc.

hidmic commented:
> I'm not sure if we can also take advantage of some of the existent tools to e.g. launch a new process, capture process stdout, etc.

Maybe, to some extent. We can probably do better, but it's not a bad start.

ivanpauno commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,258 @@
+            'or a launch description, locals pair'
+        )
+        ld = ret
+        if isinstance(ret, tuple):
```
I'm planning to delete the option of return a dictionary with variables.
There's no reason to have that, we can use pytest fixtures.
Pytest fixtures guarantee to cache the value passed to other fixtures and to the test case, even when the scope is 'function'. e.g.:

```python3
@pytest.fixture
def foo():
    ...

@pytest.fixture
def bar(foo):
    ...

def test_case(foo, bar):
    # the value of `bar` we get here is the same one we got in `foo`
    ...
```
It also complicates the proposal commented [here](https://gist.github.com/ivanpauno/761a7cbca83335ea081512f3494304ba#gistcomment-3885449) for function scoped tests cases ....

@hidmic what do you think?

hidmic replied:
I think it looks cool and idiomatic. But it has two downsides: it forces the launch fixture to be split at least in two to pass things around, and it opens the door to scope issues. If I'm not mistaken, a function scoped launch fixture getting its actions from a module scoped regular fixture will fail subtlety. Unless we address https://github.com/ros2/launch/issues/113.

> It also complicates the proposal commented here for function scoped tests cases ....

How bad it gets?

ivanpauno replied:
>  If I'm not mistaken, a function scoped launch fixture getting its actions from a module scoped regular fixture will fail subtlety. Unless we address #113.

Yes, I would consider that a programmer error, but it might be hard to understand.

> How bad it gets?

I think it doesn't get that bad... I will push more changes soon.

ivanpauno replied:
resolving this comment for the moment, we can discuss this later if needed

ivanpauno commented:
@hidmic could you take another look?

I have added support for shutdown tests, including generator/asyncgenerators based ones.
The code needs more polishing, more tests, and maybe we should handle launch_service higyene in a better way; but I don't think we need many more features (except for improving tools to expect for process output/launch process/etc, but that's kind of unrelated).

ivanpauno commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,457 @@
+def pytest_collection_modifyitems(session, config, items):
+    """Reorder tests, so shutdown tests happen after the corresponding fixture teardown."""
+
+    def cmp(left, right):
+        leftm = left.get_closest_marker('launch_testing')
+        rightm = right.get_closest_marker('launch_testing')
+        if None in (leftm, rightm):
+            return 0
+        fixture = leftm.kwargs['fixture']
+        if fixture is not rightm.kwargs['fixture']:
+            return 0
+        if fixture._pytestfixturefunction.scope == 'function':
+            return 0
+        left_is_shutdown = int(left._launch_testing_is_shutdown)
+        right_is_shutdown = int(right._launch_testing_is_shutdown)
+        return left_is_shutdown - right_is_shutdown
+
+    # python sort is guaranteed to be stable
+    items.sort(key=functools.cmp_to_key(cmp))
```
I think I'm going to rewrite the sorting manually.
the currently  cmp() method doesn't respect the transivity property:

you might have a < b, b=c, c=a.
I don't think python sorting handles that correctly, I'm not sure honestly.

ivanpauno replied:
I also want to double check that all tests using the same launch test fixture are already one next to the other in the items list, I think that's the case but I'm not completely sure.
If that's not the case, we need to do something more complex here.

ivanpauno replied:
What pytest default hook does https://github.com/pytest-dev/pytest/blob/a83b359cf0902e0e55b79411dd54480be8a0aa4e/src/_pytest/fixtures.py#L1604 (which is run before this hook).

ivanpauno replied:
done, I didn't use exactly that snippet though, see https://github.com/ros2/launch/pull/528/commits/46ba3b249c5f23869d750816bba35c67f09d4ce1

hidmic commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,457 @@
+    for item in items:
+        marker = item.get_closest_marker('launch_testing')
+        if marker is not None and item._launch_testing_is_shutdown:
+            item.fixturenames.remove(marker.kwargs['fixture'].__name__)
```
Wild idea: what if we use dependent fixtures to handle launch service lifecycle instead of finalizers?

Let's say we have a launch fixture A. On decoration, we introduce two more fixtures B and C with the same scope. Both B and C depend on A. B takes the value of A and manages the launch service lifecycle: it pushes it into the event loop, yields, and on finalization it shuts down (or not) and waits for it to wrap up. C takes the value of A, runs the launch service to completion, and returns. Regular launch tests depend on B. Post-shutdown launch tests depend on A, unless there's only post-shutdown launch tests in which case they all depend on C. 

That should take care of test ordering, passing extra arguments, and the _all shutdown tests_ scenario. WDYT?

ivanpauno replied:
> @ivanpauno hmm I have the impression this won't work when all tests are shutdown tests.

Yes, this is still true.
I'm inclined to delete that option.

> Wild idea: what if we use dependent fixtures to handle launch service lifecycle instead of finalizers?

Maybe that works, I have to another look to this alternative.

hidmic replied:
Yeah, it may or may not. When I thought about it looked like it did :sweat_smile:

ivanpauno replied:
> Yeah, it may or may not. When I thought about it looked like it did sweat_smile

I gave it a try, it's impossible to support parameterized fixtures in that way.
(maybe it's possible, but pytest doesn't seem to have been think to do that kind of trick)

ivanpauno replied:
we discussed this offline, we agreed the current approach seems fine.

hidmic reviewed (commented):
Alright, I left some comments, questions, suggestions, and ideas. This is complex code, sorry if I mixed things up :sweat_smile:

ivanpauno commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,457 @@
+    def inner(**kwargs):
+        update_arguments(kwargs, args)
+        agen = func(**kwargs)
+        shutdown.agen = agen
```
yeah

ivanpauno replied:
fixed in https://github.com/ros2/launch/pull/528/commits/d8a5ef9f487ae4e5852d94e426643d1adb903a3d

ivanpauno commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,457 @@
+        ls = pyfuncitem.funcargs['launch_service']
+        on_shutdown = functools.partial(
+            finalize_launch_service, ls, eprefix=f'When running test {func.__name__}')
+        before_test = on_shutdown if shutdown_test else None
```
> meta: immediate fixture shutdown for pure post-shutdown tests? There may not even be enough time for anything meaningful to happen

True

> I'm inclined to think that such tests go hand-in-hand with self-terminating fixtures.

What do you mean?

hidmic replied:
Oh, I never replied. I mean that a pure-shutdown test would only make sense with a fixture that terminates itself (by going idle with `shutdown_on_idle=True`, by emitting a `Shutdown()` event) after some time. If the fixture doesn't terminate itself, what will? when?

ivanpauno replied:
> Oh, I never replied. I mean that a pure-shutdown test would only make sense with a fixture that terminates itself (by going idle with shutdown_on_idle=True, by emitting a Shutdown() event) after some time. If the fixture doesn't terminate itself, what will? when?

The fixture teardown step or the shutdown tests, both are currently shutting down the launc service.
The only issue is that nothing will setup the launch service, because pure shutdown tests don't depend on the launch description fixture.

ivanpauno replied:
> The only issue is that nothing will setup the launch service, because pure shutdown tests don't depend on the launch description fixture.

this is now fixed, in a pure shutdown test we're now launching the fixture and waiting to shutdown to run it

ivanpauno commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,457 @@
+
+    @functools.wraps(func)
+    def shutdown(**kwargs):
+        gen = getattr(shutdown, 'gen', None)
```
Thanks for the suggestion, fixed in https://github.com/ros2/launch/pull/528/commits/2ef726df26959a1294b72ff90818cb7452dd709b.

clalancette commented:
> > On a related note, would it make sense to roll out a separate launch_pytest package? Imports would be shorter and it's easier to draw a line between this new thing and launch_testing. WDYT @ivanpauno @wjwwood @jacobperron ?
> 
> The imports wouldn't really be much shorter.
> Having a separate package sounds fine to me, I actually started that way and then integrated the changes in launch_testing.

On the one hand, replacing `launch_testing` completely with the new pytest-based thing is where we eventually want to be.  On the other hand, just replacing all of the guts of `launch_testing` is probably going to break all downstream packages that have managed to use `launch_testing` successfully.

Because of that, my inclination is to make a new package, and once that is in have `import launch_testing` raise a DeprecationWarning in Humble.  Then for I-Turtle we can remove the old `launch_testing`.  What do you think about that plan?

On a different meta-note, I'd also like to see some examples on how to use this new functionality in downstream packages.  For instance, how would I change https://github.com/ros-teleop/teleop_tools/blob/foxy-devel/joy_teleop/test/test_action_fibonacci.py to use this new package?  Maybe those examples should be in here, or maybe they should be over in https://github.com/ros2/examples, or maybe they should be in the documentation.

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2021-9-16/22372/1

ivanpauno commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,457 @@
+                args[name] = value
+        if inspect.iscoroutinefunction(func):
+            pyfuncitem.obj = wrap_coroutine(func, args, event_loop, before_test)
+        elif inspect.isgeneratorfunction(func):
```
Not really, I think the shutdown flag is being ignored in that case, though maybe better we should show an error.
I'm also having problems to support pure post-shutdown tests, so maybe I'll delete that option altogether.

ivanpauno replied:
See https://github.com/ros2/launch/pull/528/commits/bf8c790fe9b3b5270884d2c5562ef0468b99852c

wjwwood commented:
> On the one hand, replacing launch_testing completely with the new pytest-based thing is where we eventually want to be. On the other hand, just replacing all of the guts of launch_testing is probably going to break all downstream packages that have managed to use launch_testing successfully.

I'm fine with this approach, to be honest. It also opens us up to being able to do `launch_next_new_cool_testing_thing` down the line when that comes around.

ivanpauno commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,457 @@
+                    wrap_asyncgen(func, args, event_loop, on_shutdown)
+                )
+            else:
+                pyfuncitem.obj = wrap_asyncgen_fscope(func, args, event_loop, on_shutdown)
```
The main issues is the async generator.
You must complete running the generator using the same event loop, thus we need only one test item.

The other way to solve the issue is to just have one `event_loop` module scoped fixture.
There's really no reason to be destroying and creating event loops, it only adds a little extra "isolation" between the test cases (which shouldn't be a problem if we are always running things to completion).

ivanpauno replied:
we discussed this offline, we agreed that the current aproach seems fine

hidmic commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,528 @@
+        next(gen)
+    except StopIteration:  # Only one ReadyToTest action must be found
+        return ready_action
+    raise RuntimeError(
+        'launch_pytest fixtures must return a LaunchDescription '
+        'containing only one ReadyToTest action'
+    )
```
@ivanpauno meta: just like `launch_testing`, this cannot really detect all `ReadyToTest` instances (thinking of includes). We need to document the fact that we'll simply pick up the first one found we can statically find. In which case, I'm slightly inclined to make this a warning. Thoughts?

hidmic commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,528 @@
+    kwargs = item.get_closest_marker('launch_testing').kwargs
+    ret = 'fixture' in kwargs and kwargs['fixture'] is not None
+    if not ret:
+        msg = (
+                    '"fixture" keyword argument is required in a pytest.mark.launch_testing() '
+                    f'decorator')
```
@ivanpauno formatting is off here.

hidmic commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,528 @@
+
+
+# Adapted from https://github.com/pytest-dev/pytest-asyncio,
+# see their license https://github.com/pytest-dev/pytest-asyncio/blob/master/LICENSE.
```
@ivanpauno meta: not sure how do we go about acknowledging code brought from elsewhere. @tfoote @clalancette ?

clalancette replied:
A few things I would think about:

1. Licensing.  Since the license on pytest-asyncio agrees 100% with this package (Apache 2.0), that works fine.
2. Copyright name.  Typically I would add the Copyright name from the headers in the `pytest-asyncio` repository, but they don't have any.  Since they don't have any attribution there, I don't think we particularly need it.
3. Attribution.  I think attributing it like this is reasonable, though I might point to a specific file at a specific hash rather than just generically at the repository.

hidmic commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,528 @@
+                print(f'moving {item.name} after {other.name}: {i} {i+j+1}')
+        if new_position is not None:
+            items.insert(new_position, items.pop(i))
+            print([item.name for item in items])
```
@ivanpauno nit: drop debugging print

ivanpauno replied:
done, https://github.com/ros2/launch/pull/528/commits/4c9cf98c3788d975e88e54bfb5c06ea6e647cb37

hidmic commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,528 @@
+                and is_shutdown_test(item) and not is_shutdown_test(other)
+            ):
+                new_position = i + 1 + j
+                print(f'moving {item.name} after {other.name}: {i} {i+j+1}')
```
@ivanpauno nit: drop debugging print

ivanpauno replied:
done in https://github.com/ros2/launch/commit/4c9cf98c3788d975e88e54bfb5c06ea6e647cb37

hidmic commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,528 @@
+        # This algo has worst case Nitems * Nlaunchtestitems time complexity.
+        # We could probably find something better, but it's not that easy because:
+        # - Tests using the same launch fixture might not be already grouped,
+        #   i.e. there might be a test not using the fixture in the middle.
```
@ivanpauno meta: if this is true, what prevents a module scope launch fixture and a function scope launch fixture from running concurrently?

ivanpauno replied:
IIUC, it's possible to run into that issue if you have weird dependencies between fixtures.

hidmic commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,528 @@
+            return items
+
+
+def get_error_context_from_obj(obj):
```
@ivanpauno nit: unused?

hidmic commented on launch_testing/test/launch_testing/new_hooks/test_function_scope.py:
```diff
@@ -0,0 +1,82 @@
+    ])
+
+
+# TODO(ivanpauno)
+# We cannot get variables from the dictionary returned by launch_description
+# because we're removing the fixture before running the tests.
+# Maybe we can delete this feature, and use generators/asyncgens.
+# We can always get the variables returned by the launch_testing fixture there.
```
@ivanpauno meta: this is no longer true, is it?

ivanpauno replied:
exactly, comment deleted in https://github.com/ros2/launch/pull/528/commits/dda4cd9f3e48a7d30d44c8dae79d59dd63c864af.

hidmic commented on launch_testing/test/launch_testing/new_hooks/test_function_scope.py:
```diff
@@ -0,0 +1,82 @@
+    return launch.LaunchDescription([
+        launch_testing.util.KeepAliveProc(),
+        launch_testing.actions.ReadyToTest(),
+    ])
```
@ivanpauno It'd be nice to test a `launch_description` that yields (though I don't see why it wouldn't work)

hidmic commented on launch_testing/launch_testing/util/__init__.py:
```diff
@@ -30,9 +30,12 @@ def KeepAliveProc():
 try:
     while True:
-        pass
+        time.sleep(1)
```
:see_no_evil:

hidmic commented on launch_testing/launch_testing/pytest/plugin.py:
```diff
@@ -0,0 +1,528 @@
+    rfn = get_launch_test_fixture(right_item)
+    if None in (lfn, rfn):
+        return False
+    if lfn != rfn:
```
@ivanpauno nit: should we check for identity?  
```suggestion
    if lfn is not rfn:
```

hidmic reviewed (commented):
Second (or third?) pass!

ivanpauno commented:
The reason why the Rpr checker fails is interesting.

It seems that we're using the pip installed pytest in the buildfarm, due to `pytest-rerunfailures` [being installed from pip](https://github.com/ros-infrastructure/ros_buildfarm/blob/833969e9f9a4e907059c4377472ae0d018f2c884/ros_buildfarm/templates/devel/devel_task.Dockerfile.em#L67) (no debian package available in Focal).
`pytest-asyncio` from `python3-pytest-asyncio` (via rosdep), and the version available in ubuntu debians is not compatible with the pip installed pytest.

I'm not using much of `pytest-asyncio`, only a fixture that I can duplicate in our plugin, but the issue is a bit unfortunate.

ivanpauno commented:
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15748)](http://ci.ros2.org/job/ci_windows/15748/)

ivanpauno commented:
@hidmic could you take another look?
I have fixed the windows issues mentioned above and I've also introduced compatibility with the version of pytest shipped in Focal, see https://github.com/ros2/launch/pull/528/commits/a9300bf71bf807b5eccaa350dbc94aa0ec3aa255 (we're using the pip install version in CI).

ivanpauno commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=15538)](http://ci.ros2.org/job/ci_linux/15538/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=10223)](http://ci.ros2.org/job/ci_linux-aarch64/10223/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=13207)](http://ci.ros2.org/job/ci_osx/13207/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=15751)](http://ci.ros2.org/job/ci_windows/15751/)

hidmic commented on launch_pytest/test/launch_pytest/tools/test_process.py:
```diff
@@ -0,0 +1,73 @@
+@pytest.mark.launch(fixture=launch_description)
+async def test_async_process_tools(dut, launch_context):
+    await tools.wait_for_start(launch_context, dut, timeout=10)
+    def check_output(output): assert output == f'hello{os.linesep}'
```
@ivanpauno meta: using `str.splitlines()` would take care of all newline nuances.

ivanpauno replied:
Fixed in https://github.com/ros2/launch/pull/528/commits/cf66f4b8cc5515c00cb5cd0a411c92edaa093647

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2021-10-28/22947/1

Blast545 commented:
@ivanpauno Can I ask you to take a look to this PR? it introduced some regressions to the repeated jobs of the buildfarm:

[nightly_linux_repeated#2459](https://ci.ros2.org/job/nightly_linux_repeated/2459/)
[nightly_linux-aarch64_repeated#1766/](https://ci.ros2.org/job/nightly_linux-aarch64_repeated/1766/)
[nightly_win_rep#2416](https://ci.ros2.org/job/nightly_win_rep/2416/)
[nightly_osx_repeated#2526](https://ci.ros2.org/job/nightly_osx_repeated/2526/)

I imagine it's missing a cleanup that's only showing once the tests are run in a sequence.

ivanpauno commented:
> @ivanpauno Can I ask you to take a look to this PR? it introduced some regressions to the repeated jobs of the buildfarm:

Interesting... I will take a look

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2021-11-18/23209/1

wep21 commented:
@ivanpauno @hidmic Is backporting `launch_pytest` to `galactic` acceptable? If so, I would like to create a backport PR for `launch_pytest`.

ivanpauno commented:
> @ivanpauno @hidmic Is backporting launch_pytest to galactic acceptable? If so, I would like to create a backport PR for launch_pytest.

I think that up to now it hasn't been used enough to say it's worth backporting.
I would rather start using it more and prove its usefulness before backporting it.
